### PR TITLE
Bruno/document local testing

### PR DESCRIPTION
**Pull request by BrunoReboul**
Facilitate constraints and templates troubleshooting by documenting and tooling how to run local test without deploying to Forseti. Benefits:

- Speedup troubleshooting
- Reduced complexity by segregating rego code issues from Forseti issues

### CODE REVIEW

File: scripts/dumps2json.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+        assetsList.append(asset)
+        if asset['ancestry_path'] == "":
+            print(f"Warning: missing ancestry path {asset['name']}")
+
+    localJSONFile = open(outputFolderPath + '/' + outputFileName, 'w')
+    localJSONFile.write(json.dumps(assetsList, indent=2))
+    localJSONFile.close()
+
+
+def get_ancestry_path(assets):
```
Comment (ocervell) [inline scripts/dumps2json.py]: This function can be sped up 3X by merging loops together:

```
def get_ancestry_path(assets):
    
    # First flush, get the root asset
    for _, asset in assets.items():
        if asset['asset_type'] == "cloudresourcemanager.googleapis.com/Organization":
            organizationPath = asset['resource']['data']['name'].replace(
                'organizations', 'organization')
            asset['ancestry_path'] = organizationPath

    if organizationPath:
        print(f"Found the organization for this dumps: {organizationPath}")
        knownProjects = {}

        for _, asset in assets.items():

            # Fix assets without resource object
            if not 'resource' in asset:
                asset['ancestry_path'] = organizationPath
                print(
                    f"warning: resource object is missing, ancestry_path set to org_path for {asset['name']}")

            # Compute the ancestry path if missing
            if not 'ancestry_path' in asset:
                asset['ancestry_path'] = build_path(
                    name=asset['name'], assets=assets)

            # Find known projects
            if 'ancestry_path' in asset and asset['asset_type'] == "cloudresourcemanager.googleapis.com/Project":
                if not asset['resource']['data']['name'] in knownProjects:
                    knownProjects[asset['resource']['data']
                                  ['name']] = asset['ancestry_path']

            # If ancestry path is empty, fix it as possible
            if asset['ancestry_path'] == "":
                assetNameBreakdown = asset['name'].split('/')
                if assetNameBreakdown[3] == 'projects':
                    if assetNameBreakdown[4] in knownProjects:
                        asset['ancestry_path'] = knownProjects[assetNameBreakdown[4]]
                    else:
                        asset['ancestry_path'] = organizationPath
                elif assetNameBreakdown[3] == 'billingAccounts':
                    asset['ancestry_path'] = 'organization/'
                else:
                    asset['ancestry_path'] = organizationPath
                #print(f"warning fixed ancestry path for: {asset['name']}")
    return assets
```

File: scripts/dumps2json.py [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+                    if assetNameBreakdown[4] in knownProjects:
+                        asset['ancestry_path'] = knownProjects[assetNameBreakdown[4]]
+                    else:
+                        asset['ancestry_path'] = organizationPath
+                elif assetNameBreakdown[3] == 'billingAccounts':
+                    asset['ancestry_path'] = 'organization/'
+                else:
+                    asset['ancestry_path'] = organizationPath
+                #print(f"warning fixed ancestry path for: {asset['name']}")
+    return assets
```
Comment (ocervell) [inline scripts/dumps2json.py]: We should be able to rewrite the method as below, in order to gain one loop, and use a dict comprehension to speed things up for the `knownProjects` var (needs testing):

```py
def get_ancestry_path(assets):

    # First flush, get the organization path
    organizationPath = None
    for _, asset in assets.items():
        if asset['asset_type'] == "cloudresourcemanager.googleapis.com/Organization":
            organizationPath = asset['resource']['data']['name'].replace(
                'organizations', 'organization')
            asset['ancestry_path'] = organizationPath

    if not organizationPath:
        print(f"Could not find the organization for this dumps")
        return assets

    # Second flush: add 'ancestry_path' field to all assets
    for _, asset in assets.items():

        # Fix assets without resource object
        if not 'resource' in asset:
            asset['ancestry_path'] = organizationPath
            print(
                f"warning: resource object is missing, ancestry_path set to org_path for {asset['name']}")

        # Compute the ancestry path if missing
        if not 'ancestry_path' in asset:
            asset['ancestry_path'] = build_path(
                name=asset['name'],
                assets=assets)

    # Third flush: Get map project name / ancestry for all projects
    knownProjects = {
        asset['resource']['data']['name']: asset['ancestry_path']
        for asset in assets
        if 'ancestry_path' in asset and
            asset['asset_type'] = "cloudresourcemanager.googleapis.com/Project"
    }

    # Fourth flush: loop through all assets and set 'ancestry_path' field if missing
    for _, asset in assets.items():
        if not asset['ancestry_path']:
            assetNameBreakdown = asset['name'].split('/')
            if assetNameBreakdown[3] == 'projects':
                if assetNameBreakdown[4] in knownProjects:
                    asset['ancestry_path'] = knownProjects[assetNameBreakdown[4]]
                else:
                    asset['ancestry_path'] = organizationPath
            elif assetNameBreakdown[3] == 'billingAccounts':
                asset['ancestry_path'] = 'organization/'
            else:
                asset['ancestry_path'] = organizationPath

    return assets
```

File: test/audit/audit.rego [hunk 3]
```diff
@@ -0,0 +1,61 @@
+
```
Comment (ocervell) [inline test/audit/audit.rego]: is there a way we can use the `audit.rego` without needing a full copy of the file under the `test/audit` folder ?

File: docs/constraint_template_local_testing.md [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+```bash
+.                         # your config-validator repo root
+├─ test                   # test folder, may be named differently
+│  ├─ assets              # this folder is OPA's data.assets
+│  │  └─ data.json        # the one json file from previous step
+│  ├─ audit               # folder for rego files supporting test
+│  │  └─ audit.rego       # policy module similar to forseti audit.go  
+│  └─ constraints         # this folder is OPA's data.constraints
+│     ├─ constraintname   # folder for the constraint to be tested
```
Comment (ocervell) [inline docs/constraint_template_local_testing.md]: `constraintname` --> `<CONSTRAINT_NAME>` to fit Google standards

File: docs/constraint_template_local_testing.md [hunk 5]
```diff
@@ -0,0 +1,136 @@
+# Test config-validator constraints and templates locally without Forseti
+
+**DISCLAIMER**: the file `data.json` used here containts all organization's assets metadata. Do NOT publish this information on a public repo. Use
+
```
[the remark this answers is not in the document]
Comment (BrunoReboul) [reply docs/constraint_template_local_testing.md]: fixed

File: docs/constraint_template_local_testing.md [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+This document is for users who wish to test constraint and
+template on their organization metadata, from their system, before deploying to forseti. It both apply to existing templates and custom templates.
```
[the remark this answers is not in the document]
Comment (BrunoReboul) [reply docs/constraint_template_local_testing.md]: done

File: docs/constraint_template_local_testing.md [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+
+**Benefits**:
+
+- Save time
```
[the remark this answers is not in the document]
Comment (BrunoReboul) [reply docs/constraint_template_local_testing.md]: done

File: docs/constraint_template_local_testing.md [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+- Reduce test complexity
```
[the remark this answers is not in the document]
Comment (BrunoReboul) [reply docs/constraint_template_local_testing.md]: done

File: docs/constraint_template_local_testing.md [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+- [Run test locally](Run_test_locally)
+- [Contact info](Contact_info)
+
+## Prerequisites
+
+- [OPA is installed](https://www.openpolicyagent.org/docs/latest/get-started/#prerequisites) locally, and added to the PATH
+
+## Get organization's assets metadata in one `data.json` file
+
+- Copy organization's CAI `resource` dump and organization's CAI `iam_policy` dump in one folder
```
[the remark this answers is not in the document]
Comment (BrunoReboul) [reply docs/constraint_template_local_testing.md]: the link already exist on line 36

File: docs/constraint_template_local_testing.md [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+  - Depending on permissions available on the organization:
+    - Donwload the 2 dumps files from Forseti CAI bucket
+    - or use [CAI directly](https://cloud.google.com/resource-manager/docs/cloud-asset-inventory/quickstart-cloud-asset-inventory) to export dumps
+      - Key point: Ensure consistency by having the two dumps exported at the same time
+- Run the python script `dump2json.py`
+  - E.g. `python3 scripts/dump2json.py ----dump-folder-path dumpsFolderName`
```
[the remark this answers is not in the document]
Comment (BrunoReboul) [reply docs/constraint_template_local_testing.md]: done

File: docs/constraint_template_local_testing.md [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+
+- [OPA is installed](https://www.openpolicyagent.org/docs/latest/get-started/#prerequisites) locally, and added to the PATH
+
+## Get organization's assets metadata in one `data.json` file
+
+- Copy organization's CAI `resource` dump and organization's CAI `iam_policy` dump in one folder
+  - Depending on permissions available on the organization:
+    - Donwload the 2 dumps files from Forseti CAI bucket
+    - or use [CAI directly](https://cloud.google.com/resource-manager/docs/cloud-asset-inventory/quickstart-cloud-asset-inventory) to export dumps
+      - Key point: Ensure consistency by having the two dumps exported at the same time
```
[the remark this answers is not in the document]
Comment (BrunoReboul) [reply docs/constraint_template_local_testing.md]: re formulated

File: docs/constraint_template_local_testing.md [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+- Run the python script `dump2json.py`
+  - E.g. `python3 scripts/dump2json.py ----dump-folder-path dumpsFolderName`
+  - Check the script's messages
+  - `data.json` file is delivered in the dumps folder
+  - `data.json` is an array of all organization's objects. Each object is a GCP asset. Each asset object contains a resource object and if applicable a iam_policy object too.
+  - `ancestry_path` used in Forseti to specify where the constraint is to be applied has but computed and added.
+  
+Example
```
[the remark this answers is not in the document]
Comment (BrunoReboul) [reply docs/constraint_template_local_testing.md]: done

File: docs/constraint_template_local_testing.md [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+│  ├─ audit               # folder for rego files supporting test
+│  │  └─ audit.rego       # policy module similar to forseti audit.go  
+│  └─ constraints         # this folder is OPA's data.constraints
+│     ├─ constraintname   # folder for the constraint to be tested
+│     │   └─ anyname.yaml # constraint ymal file
+│     └─ ...              # additional constraint folders if needed
+└─ ...
+```
+
+The constraint yaml file to be tested is copied from `policies/constraints` to `test/constraints/constraintname/`
```
[the remark this answers is not in the document]
Comment (BrunoReboul) [reply docs/constraint_template_local_testing.md]: we do not need to have the constraint name here. anyname will fit. only the folder name matter.
replaced by <ANY_NAME>.yaml

File: docs/constraint_template_local_testing.md [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+
+`opa run lib/ validator/ test/iamAuditLog_orgOnly_iap/`  
+
+## Run test locally
+
+- Launch [OPA REPL](https://www.openpolicyagent.org/docs/latest/get-started#goals) by executing `opa run lib/ validator/ test/`
+- From OPA command line
+  - `package validator.gcp.lib` this should produce no output
+  - `audit` this execute the audit rego rule and outputs found non compliances
+  - Execution time depens on assets number and constraints complexity
```
[the remark this answers is not in the document]
Comment (BrunoReboul) [reply docs/constraint_template_local_testing.md]: done

File: scripts/dumps2json.py [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+import argparse
+import sys
+import os
+import glob
+import json
+
+outputFileName = 'data.json'
+dataKeysToRemove = ['oauth2ClientId']
+dataKeysToRedact = ['serverCaCert']
+replaceInfoWith = '*****'
```
[the remark this answers is not in the document]
Comment (BrunoReboul) [reply scripts/dumps2json.py]: No requirement from Google found here: [http://google.github.io/styleguide/pyguide.html](http://google.github.io/styleguide/pyguide.html)

About PEP:
- Functions are already named in snake_case.
- Please confirm if variable must be too ?

File: scripts/dumps2json.py [hunk 16]
```diff
@@ -0,0 +1,176 @@
...
+
+
+def parse_args(argv):
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--dump-folder-path', required=True, dest='dumpFolderPath',
```
[the remark this answers is not in the document]
Comment (BrunoReboul) [reply scripts/dumps2json.py]: see previous

File: scripts/dumps2json.py [hunk 17]
```diff
@@ -0,0 +1,176 @@
...
+                        help='path to the folder containing the cai dump files one for resources, one for iam policies')
+    parser.add_argument('--output-folder-path', required=True, dest='outputFolderPath',
```
[the remark this answers is not in the document]
Comment (BrunoReboul) [reply scripts/dumps2json.py]: see previous

File: scripts/dumps2json.py [hunk 18]
```diff
@@ -0,0 +1,176 @@
...
+def get_ancestry_path(assets):
+    # First flush, get the root asset
+    for _, asset in assets.items():
+        if asset['asset_type'] == "cloudresourcemanager.googleapis.com/Organization":
+            organizationPath = asset['resource']['data']['name'].replace(
+                'organizations', 'organization')
+            asset['ancestry_path'] = organizationPath
+
+    if organizationPath:
+        print(f"Found the organization for this dumps: {organizationPath}")
```
[the remark this answers is not in the document]
Comment (BrunoReboul) [reply scripts/dumps2json.py]: CAI export dump for one organiztion (or less, folder, projet).
This script is to be used only with org dumps to perform a consistent work on ancestry_path

File: scripts/dumps2json.py [hunk 19]
```diff
@@ -0,0 +1,176 @@
...
+        # Second flush, fix assets without resource object
+        for _, asset in assets.items():
+            if not 'resource' in asset:
+                asset['ancestry_path'] = organizationPath
+                print(
+                    f"warning: resource object is missing, ancestry_path set to org_path for {asset['name']}")
+
+        # Thrid flush, compute the ancestry path when possible
```
[the remark this answers is not in the document]
Comment (BrunoReboul) [reply scripts/dumps2json.py]: done

File: test/audit/audit.rego (hunk 3 above)
Comment (BrunoReboul) [reply test/audit/audit.rego]: yes it works too, but take care to still have `ancestry_path` populated.
Files like in `validator/test/fixture` wont work as the `ancestry_path` is missing

File: scripts/dumps2json.py (hunk 2 above)
Comment (BrunoReboul) [reply scripts/dumps2json.py]: TBD

File: scripts/dumps2json.py [hunk 20]
```diff
@@ -0,0 +1,176 @@
...
+        for _, asset in assets.items():
+            if not 'ancestry_path' in asset:
+                asset['ancestry_path'] = build_path(
+                    name=asset['name'], assets=assets)
```
[the remark this answers is not in the document]
Comment (BrunoReboul) [reply scripts/dumps2json.py]: TBD

File: scripts/dumps2json.py [hunk 21]
```diff
@@ -0,0 +1,176 @@
...
+def main(dumpFolderPath, outputFolderPath):
+    assetsDict = {}
+    assetsList = []
+    for dumpFilename in glob.glob(os.path.join(dumpFolderPath, '*.dump')):
+        with open(dumpFilename, 'r') as caiDumpFile:
+            numberOfAssets = sum(1 for _ in caiDumpFile)
+            print(f'number of asset: {numberOfAssets} in {dumpFilename}')
+        with open(dumpFilename, 'r') as caiDumpFile:
+            i = 0
+            for line in caiDumpFile:
```
[the remark this answers is not in the document]
Comment (BrunoReboul) [reply scripts/dumps2json.py]: done

**Comment by ocervell** [comment]
@morgante do we have docs somewhere for local config-validator testing (without Forseti) ?

File: scripts/dumps2json.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (ocervell) [reply scripts/dumps2json.py]: From the above link I've shared:

```
Variable names follow the same convention as function names.

mixedCase is allowed only in contexts where that's already the prevailing style (e.g. threading.py), to retain backwards compatibility.
```

In the link you've shared, it's also specified: http://google.github.io/styleguide/pyguide.html#3164-guidelines-derived-from-guidos-recommendations

**Comment by morgante** [comment]
@ocervell I'm not quite sure I follow though, because constraint template testing should have no need to look at ancestry info. In fact, constraint template authors should never be inspecting the `ancestry_path` field at all.

I'd like to understand what part of `make test` isn't working today for template authorship/debugging.

As for debugging a customer's actual environment, that is best done via go/cft-scorecard.

I do agree the conversion from `.dump` to `.json` is potentially useful. Instead of making a whole new script/testing approach, I'd instead suggest we significantly trim down this script to make it a utility for populating test fixtures within the existing testing system/managing that. I don't think we should have two different testing approaches in 1 repo.

**Comment by morgante** [comment]
@ocervell Got it, I think scorecard should be able to provide that flow.

**Pull request closed without merging**